### ISSUE TITLE
Parallel Percy tests treated as separated builds on CircleCI

### ISSUE THREAD

**Issue by zeptonaut** [percy/cli#355]
Since migrating to @percy/cypress 3 (the first version of @percy/cypress that uses @percy/cli), parallel runs have started to be treated as separate Percy builds.

The user-facing result of this is that almost every PR reports a Percy diff that needs approval because the set of screenshots treated as the canonical ones are whatever shard finished first on the latest `master` build, which is likely to not be the same set of screenshots from the shard that finished first on the current branch build.

We run our tests in CircleCI on 5 separate shards. After a single CI run of a new PR, the following appeared in the Percy dashboard:

![image](https://user-images.githubusercontent.com/1407397/119525162-4653e700-bd4c-11eb-8465-e2121a15339d.png)

Up until this migration, parallel tests "just worked"  in the way suggested by this documentation: https://docs.percy.io/docs/parallel-test-suites. Based on [this test](https://github.com/percy/cli/blob/76ec62a8760c712cae9118a4dd2a4a6eb49e5214/packages/env/test/circle.test.js), it certainly looks like the intention is for parallel Percy tests to continue to "just work" on CircleCI. However, it looks like they don't.

A few potential causes for what's going on:
- `PERCY_PARALLEL_TOTAL` isn't being set to -1
- `PERCY_PARALLEL_NONCE` isn't correctly being set to the CircleCI build number
- ~~The user now needs to supply the `--parallel` flag to percy-cli, which they didn't before @percy/cli.~~ 
**Update**: I tried supplying the `--parallel` flag and it had no effect.
- ???

I'm going to continue to debug this, but thought I'd file an issue so that the team is aware this is a problem.

If it's helpful for debugging, the Percy build number for one of the affected runs is 10623605.

**EDIT**: If it's helpful, the command that we use to run our Percy tests in parallel is:

```sh
$ circleci tests glob "cypress/integration/**/*.js" | circleci tests split --split-by=timings > ../test-results/cypress_tests.txt
$ yarn run percy exec --parallel -- cypress run --reporter junit --reporter-options "mochaFile=../test-results/cypress/results-[hash].xml" --spec $(cat ../test-results/cypress_tests.txt)
```

**Comment by zeptonaut** [comment]
Update: tried setting `PERCY_PARALLEL_TOTAL to -1, but it had no effect:

![image](https://user-images.githubusercontent.com/1407397/119531004-88335c00-bd51-11eb-9622-10197fab0986.png)

**Comment by zeptonaut** [comment]
Thanks for the response! It looks like you were right on all three accounts:

- `yarn run` was incorrectly forwarding the `--` in the command being run
- `--parallel` should have fixed the problem
- `PERCY_PARALLEL_TOTAL=-1` should have fixed the problem

# Ultimate solution
The ultimate solution here was to move my "run Cypress with Percy" into its own command in my package.json:

`package.json`:
```json
{
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "//": "...",
    "cypress:percy-ci-run": "percy exec --parallel -- cypress run",
  },
```

`.circleci/config.yaml`:
```yaml
circleci tests glob "cypress/integration/**/*.js" | circleci tests split --split-by=timings > ../test-results/cypress_tests.txt
yarn run cypress:percy-ci-run --record --reporter junit --reporter-options "mochaFile=../test-results/cypress/results-[hash].xml" --spec $(cat ../test-results/cypress_tests.txt)
```

By putting the `percy exec --` in a command and then calling that command from the YAML file instead of calling `yarn run` directly from the YAML file, we avoid any problems with `yarn` forwarding `--` in `yarn run` incorrectly.

# Why didn't `--parallel` work?
As you mentioned above, it was being incorrectly forwarded from `yarn run` to the `percy exec` command. After moving `--parallel` into an command and calling that command instead of using `yarn run`, things worked correctly.

# Why didn't setting `PERCY_PARALLEL_TOTAL=-1` work? 
It turned out I was setting this on the wrong Docker container layer. In my `.circleci/config.json`, I was setting the environment variable with:

```yaml
  run_cypress_tests:
    docker:
      - image: canopychannels/ci:v8
        auth:
          username: canopychannels
          password: $DOCKERHUB_PASSWORD
        environment:
          PGHOST: 127.0.0.1
          PGUSER: root
          # Use system Chromium for Puppeteer instead of Percy's Chromium.
          # This decreases the size of our node_modules/ cache by ~500MB.
          PUPPETEER_SKIP_CHROMIUM_DOWNLOAD: true
          PUPPETEER_EXECUTABLE_PATH: /usr/bin/google-chrome
      - image: circleci/postgres:12.1-alpine
        environment:
          POSTGRES_USER: root
          POSTGRES_DB: circle-test_test
          PERCY_PARALLEL_TOTAL: -1
```

whereas I needed to be setting the variable on the topmost layer:

```yaml
  run_cypress_tests:
    docker:
      - image: canopychannels/ci:v8
        auth:
          username: canopychannels
          password: $DOCKERHUB_PASSWORD
        environment:
          PGHOST: 127.0.0.1
          PGUSER: root
          # Use system Chromium for Puppeteer instead of Percy's Chromium.
          # This decreases the size of our node_modules/ cache by ~500MB.
          PUPPETEER_SKIP_CHROMIUM_DOWNLOAD: true
          PUPPETEER_EXECUTABLE_PATH: /usr/bin/google-chrome
          PERCY_PARALLEL_TOTAL: -1
      - image: circleci/postgres:12.1-alpine
        environment:
          POSTGRES_USER: root
          POSTGRES_DB: circle-test_test
```

# How could this problem have been avoided?
Above, you mentioned that:

> By default Percy CLI no longer automatically sets the parallel total based on CI container count. In the past this put way too many people's builds into "parallel mode" when they really were running something like lint & tests along side each other.

I'm fairly sure that this transition happened when going from @percy/agent to @percy/cli (and the major version bump of @percy/cypress from 2 to 3). In our case, this was a breaking change, which is fine: this is a major version bump, after all.

However, the default behavior change from "parallelism works by default" to "parallelism needs an extra parameter to work" probably should have been included in the migration guide. I did the migration myself and read through the guide and would have easily made the config change in our CircleCI. Instead, my team ended up with 1+ months of Percy in a broken state until I spent the 2 or 3 hours to investigate the problem more deeply, which all feels like it could have been avoided. (For what it's worth, things had gotten so bad that I actually filed, but ultimately closed, an issue titled "Remove Percy from our continuous integration".)

Really appreciate the help figuring this out @Robdel12! Thanks for the quick response.
